Move SignalR to peerDependencies

maciejewiczow opened the pull request:
- Move SignalR to peerDependencies
- Add redux as peerDependency
- Add script that installs peerDependencies automatically in package development mode
- Remove duplicate redux types from `types` file
- Add type arguments to `withCallbacks`, to fix a typescript error that popped up

maciejewiczow committed:
Move signalr to peerDeps, add redux as peerDep
 * Add script to install peerDeps in package development
 * Remove signalr re-export

maciejewiczow committed:
Remove redux types from types file

maciejewiczow committed:
Add constraints to dispatch generic type

maciejewiczow committed:
Add generic types to withCallback

maciejewiczow commented:
I also think that type arguments order should be reversed:
```diff
- interface WithCallbacks<D extends Dispatch = Dispatch, S = any>
+ interface WithCallback<S = any, D extends Dispatch = Dispatch>
```
Imo that would be better, because you rarely will provide your own dispatch when using redux, but state almost always needs to be custom. If dispatch is first, then you always need to pass in Dispatch when you want to provide a custom state type, even if you don't actually need to change the default dispatch type.
I did not include this suggestion in this PR, because idk if it wouldn't have been too big of a change.

Omhet commented:
Thank you for your help!

Could you please explain what's the point of the install-peers dependency?

Omhet commented:
Are these changes backward-compatible? I'm just wondering how should I update the version, is it just a minor update or a major one

Omhet commented:
Yes, I agree that State type should come first, idk why I made it last. Well, I think we can add such changes and do a major update, so no current user will have any difficulties

maciejewiczow commented:
> Thank you for your help!
> 
> Could you please explain what's the point of the install-peers dependency?

I added it because when I moved signalr to peerDependencies, it wasn't being installed anymore when running `npm install`. It caused issues when working on the package, because typescript was complaining that a package is missing. When I tried to add it manually it was being registered as a dependency again. So this instal-peers package installs peerDependencies for development anytime `npm install` is run, without adding them to `package.json` again. In that way you have the packages in development, but they are not added to the final package when publishing it (anyway that's how it's supposed to work).

maciejewiczow commented:
> Are these changes backward-compatible? I'm just wondering how should I update the version, is it just a minor update or a major one

The only thing that is not backwards compatible is that I removed the export from signalr. So if someone was importing SignalR things through this package, it will no longer work. And they will need to add SignalR to theirs project dependencies themselves now. 
And this change of type arguments order is definitely not backwards compatible, that's why I did not include this suggestion in the PR

Omhet commented:
Got it. Thanks!

I think, I will change the order of types according to your suggestion and then release a new major version
